lib/cdo/contact_rollups/v2/pardot.rb

```diff
@@ -42,4 +54,115 @@ def self.retrieve_new_ids(last_id)
 
     mappings
   end
+
+  # Batch-creates prospects in Pardot.
+  # Request is only sent to Pardot when the batch is big enough.
+  # @param [Hash] contact
+  # @param [Boolean] eager_submit
+  # @return [Array] two arrays, one for all submitted prospects and one for Pardot errors
+  def self.batch_create_prospects(contact, eager_submit = false)
+    submissions = []
+    errors = []
+
+    prospect = convert_to_prospect_fields contact
+    @@prospect_batch << prospect
+
+    url = build_batch_url BATCH_CREATE_URL, @@prospect_batch
+    if url.length > URL_LENGTH_THRESHOLD || @@prospect_batch.size == MAX_PROSPECT_BATCH_SIZE || eager_submit
+      # TODO: rescue Net::ReadTimeout from submit_prospect_batch and tolerate a certain number of failures.
+      # Don't retry an insert because it will create duplicate Pardot prospects.
+      errors = submit_batch_request BATCH_CREATE_URL, @@prospect_batch
+      submissions = @@prospect_batch
+      @@prospect_batch = []
+    end
+
+    [submissions, errors]
+  end
+
+  # Batch-creates prospect in Pardot. Request is sent immediately.
+  def self.batch_create_remaining_prospects
```

Comment:
I like this (just doing last few prospects in a separate method) instead of trying to be fancy and do it all in one above.